hkmc2/shared/src/test/mlscript/lifter/StackSafetyLift.mls

```diff
@@ -24,67 +25,45 @@ fun hi(n) =
   else hi(n - 1)
 hi(0)
 //│ JS (unsanitized):
-//│ let hi;
-//│ let Cont$func$hi$1, res, $_stack$_safe$_body$_, doUnwind$, Cont$func$hi$$;
-//│ Cont$func$hi$$ = function Cont$func$hi$$(isMut, n, pc) {
-//│   let tmp, tmp1;
-//│   if (isMut === true) {
-//│     tmp = new Cont$func$hi$1(pc);
-//│   } else {
-//│     tmp = globalThis.Object.freeze(new Cont$func$hi$1(pc));
-//│   }
-//│   tmp1 = tmp(n);
-//│   return tmp1
-//│ };
-//│ (class Cont$func$hi$ extends runtime.FunctionContFrame.class {
-//│   static {
-//│     Cont$func$hi$1 = this
-//│   }
-//│   constructor(pc) {
-//│     return (n) => {
-//│       let tmp;
-//│       tmp = super(null);
-//│       this.n = n;
-//│       this.pc = pc;
-//│       return this;
-//│     }
-//│   }
-//│   #n;
-//│   get n() { return this.#n; }
-//│   set n(value) { this.#n = value; }
-//│   resume(value$) {
-//│     return hi(this.n)
-//│   }
-//│   toString() { return runtime.render(this); }
-//│   static [definitionMetadata] = ["class", "Cont$func$hi$"]; 
-//│ });
-//│ doUnwind$ = function doUnwind$(n, res1, pc) {
-//│   res1.contTrace.last.next = Cont$func$hi$$(true, n, pc);
-//│   res1.contTrace.last = res1.contTrace.last.next;
-//│   return res1
-//│ };
+//│ let hi, tmp;
+//│ let $_stack$_safe$_body$_;
 //│ hi = function hi(n) {
-//│   let scrut, tmp;
+//│   let scrut, tmp1, pc;
 //│   let stackDelayRes;
 //│   runtime.stackDepth = runtime.stackDepth + 1;
 //│   stackDelayRes = runtime.checkDepth();
 //│   if (stackDelayRes instanceof runtime.EffectSig.class) {
-//│     return doUnwind$(n, stackDelayRes, 0)
+//│     return runtime.unwind(stackDelayRes, 1, hi, null, "StackSafetyLift.mls:23:1", 0, null, 2, tmp1, n)
 //│   }
-//│   scrut = n === 0;
-//│   if (scrut === true) {
-//│     return 0
+//│   if (runtime.isResuming === true) {
+//│     let saveOffset;
+//│     pc = runtime.resumePc;
+//│     tmp1 = runtime.resumeArr.at(runtime.resumeIdx);
+//│     saveOffset = runtime.resumeIdx + 1;
+//│     n = runtime.resumeArr.at(saveOffset);
+//│     runtime.isResuming = false;
 //│   } else {
-//│     tmp = n - 1;
-//│     return hi(tmp)
+//│     pc = 0;
+//│   }
+//│   main: while (true) {
+//│     if (pc === 0) {
+//│       scrut = n === 0;
+//│       if (scrut === true) {
+//│         return 0
+//│       } else {
+//│         tmp1 = n - 1;
```

Comment:
Isn't `tmp1 = n - 1 ` going to be computed multiple times if the `hi` call below suspends and we later resume here?  I would expect resumption to jump to a state where `tmp1` is already computed (after all, the variable is restored above before being overwritten here).

Reply:
Never mind, I missed that this one was tail-recursive. But as discussed, when there's only one state, we should omit the `while` and `if (pc...)`